src/resolver.ts

```diff
@@ -2851,14 +2854,35 @@ export class Resolver extends DiagnosticEmitter {
     /** Contextual types, i.e. `T`. */
     ctxTypes: Map<string,Type> = new Map(),
     /** How to proceed with eventual diagnostics. */
-    reportMode: ReportMode = ReportMode.Report
+    reportMode: ReportMode = ReportMode.Report,
+    /** Contextual flow. */
+    ctxFlow: Flow | null = null
   ): Function | null {
     let classInstance: Class | null = null; // if an instance method
     let instanceKey = typeArguments ? typesToString(typeArguments) : "";
 
     // Instance method prototypes are pre-bound to their concrete class as their parent
     if (prototype.is(CommonFlags.Instance)) {
-      classInstance = assert(prototype.getBoundClassOrInterface());
+
+      // The actual class instance may be a subclass of the bound class
+      if (this.currentThisExpression && ctxFlow) {
+        // In the case of a function or property that uses the polymorphic `this` type,
+        // this is important, so the return is typed as the actual class instance.
+        // Note: It should work without this outer type check, and does when testing, but fails the "bootstrap" build.
+        // TODO: Figure out why and remove the extra check.
```

Comment:
Could you analyze more detail about bootstrap? What is the meaning of "outer type check"?